vpr/src/route/router_lookahead_cost_map.cpp

```diff
@@ -0,0 +1,408 @@
+#include "router_lookahead_cost_map.h"
+
+#include "router_lookahead_map_utils.h"
+#include "globals.h"
+#include "echo_files.h"
+
+#ifdef VTR_ENABLE_CAPNPROTO
+#    include "capnp/serialize.h"
+#    include "extended_map_lookahead.capnp.h"
+#    include "ndmatrix_serdes.h"
+#    include "mmap_file.h"
+#    include "serdes_utils.h"
+#endif
+
+// Distance penalties filling are calculated based on available samples, but can be adjusted with this factor.
+static constexpr float PENALTY_FACTOR = 1.f;
+static constexpr float PENALTY_MIN = 1e-12f;
+
+// also known as the L1 norm
+static int manhattan_distance(const vtr::Point<int>& a, const vtr::Point<int>& b) {
+    return abs(b.x() - a.x()) + abs(b.y() - a.y());
+}
+
+template<class T>
+static constexpr const T& clamp(const T& v, const T& lo, const T& hi) {
+    return std::min(std::max(v, lo), hi);
+}
+
+template<typename T>
+static vtr::Point<T> closest_point_in_rect(const vtr::Rect<T>& r, const vtr::Point<T>& p) {
+    if (r.empty()) {
+        return vtr::Point<T>(0, 0);
+    } else {
+        return vtr::Point<T>(clamp<T>(p.x(), r.xmin(), r.xmax() - 1),
+                             clamp<T>(p.y(), r.ymin(), r.ymax() - 1));
+    }
+}
+
+// build the segment map
+void CostMap::build_segment_map() {
+    const auto& device_ctx = g_vpr_ctx.device();
+    segment_map_.resize(device_ctx.rr_nodes.size());
+    for (size_t i = 0; i < segment_map_.size(); ++i) {
+        auto& from_node = device_ctx.rr_nodes[i];
+
+        int from_cost_index = from_node.cost_index();
+        int from_seg_index = device_ctx.rr_indexed_data[from_cost_index].seg_index;
+
+        segment_map_[i] = from_seg_index;
+    }
+}
+
+// resize internal data structures
+void CostMap::set_counts(size_t seg_count) {
+    cost_map_.clear();
+    offset_.clear();
+    penalty_.clear();
+    cost_map_.resize({1, seg_count});
+    offset_.resize({1, seg_count});
+    penalty_.resize({1, seg_count});
+    seg_count_ = seg_count;
+}
+
+// cached node -> segment map
+int CostMap::node_to_segment(int from_node_ind) const {
+    return segment_map_[from_node_ind];
+}
+
+static util::Cost_Entry penalize(const util::Cost_Entry& entry, int distance, float penalty) {
```

Comment:
I am adding a comment in the code. The basic idea though is that, if a specific `(delta_x, delta_y)` coordinates pair falls out of a segment bounding box, the returned cost is a resultof the following equation:

```
delay + distance * penalty * PENALTY_FACTOR
```

Where:
- delay: delay of the closest point in the bounding box for the specific wire segment
- distance: this can assume two values:
  - 0: in case the deltas fall within the bounding box (in this case no penalty is added)
  - non-0: in case the point is out of the bounding box, the value is the manhattan distance between the point and the closest point within the bounding box.
- penalty: value determined when building the lookahead and relative to a specific segment type (meaning there is one penalty value for each segment type). The base delay_penalty value is a calculated as follows:

```
(max_delay - min_delay) / max(1, manhattan_distance(max_location, min_location))
```

- PENALTY_FACTOR: impact of the penalty on the total delay cost.

Reply:
This didn't improve things much, if at all, from what I remember, so it might be worth removing this and all penalty calculations and testing that.

Reply:
Comment added